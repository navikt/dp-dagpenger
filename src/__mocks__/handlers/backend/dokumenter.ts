import { HttpResponse } from "msw";

export const dokumenter = ({ request }) => {
  if (!request.headers.has("Authorization")) {
    return HttpResponse.json({
      errors: [
        {
          message: "Unautorized",
        },
      ],
    });
  }

  if (!("fnr" in request.variables)) {
    return HttpResponse.json({
      errors: [
        {
          message: "Mangler fnr",
        },
      ],
    });
  }
  
  if (
    !request.headers.has("Nav-Callid") ||
    !request.headers.has("Nav-Consumer-Id")
  ) {
    throw new Error("Request må ha Nav-Call-Id og Nav-Consumer-Id");
  }

  return HttpResponse.json({
    data: sample,
  });
};

export const dokument = ({ request }) => {
  if (!request.headers.has("Authorization")) {
    return new HttpResponse(null, { status: 401 });
  }
  if (
    !request.headers.has("Nav-Callid") ||
    !request.headers.has("Nav-Consumer-Id")
  ) {
    throw new Error("Request må ha Nav-Call-Id og Nav-Consumer-Id");
  }

  return new HttpResponse("binær", {
    headers: {
      "Content-disposition": "inline",
    },
  });
};

const sample = {
  dokumentoversiktSelvbetjening: {
    journalposter: [
      {
        journalpostId: "429111291",
        tittel: "MASKERT_FELT",
        journalposttype: "U",
        journalstatus: "FERDIGSTILT",
        tema: "DAG",
        avsender: null,
        mottaker: {
          id: "123",
          tpe: "FNR",
        },
        relevanteDatoer: [
          {
            dato: "2018-01-01T12:00:00",
            datotype: "DATO_OPPRETTET",
          },
        ],
        dokumenter: [
          {
            dokumentId: "441010176",
            tittel: "MASKERT_FELT",
          },
        ],
      },
      {
        journalpostId: "429108246",
        tittel: "MASKERT_FELT",
        journalposttype: "U",
        journalstatus: "FERDIGSTILT",
        tema: "DAG",
        avsender: null,
        mottaker: {
          id: "123",
          tpe: "FNR",
        },
        relevanteDatoer: [
          {
            dato: "2018-01-01T12:00:00",
            datotype: "DATO_OPPRETTET",
          },
        ],
        dokumenter: [
          {
            dokumentInfoId: "441007131",
            tittel: "MASKERT_FELT",
          },
        ],
      },
      {
        journalpostId: "428965411",
        tittel: "MASKERT_FELT",
        journalposttype: "I",
        journalstatus: "JOURNALFOERT",
        tema: "DAG",
        avsender: {
          id: "123",
          tpe: "FNR",
        },
        mottaker: null,
        relevanteDatoer: [
          {
            dato: "2018-01-01T12:00:00",
            datotype: "DATO_OPPRETTET",
          },
        ],
        dokumenter: [
          {
            dokumentInfoId: "440831549",
            tittel: "MASKERT_FELT",
          },
          {
            dokumentInfoId: "440831548",
            tittel: "MASKERT_FELT",
          },
        ],
      },
      {
        journalpostId: "429111291",
        tittel: "MASKERT_FELT",
        journalposttype: "U",
        journalstatus: "FERDIGSTILT",
        tema: "OPP",
        avsender: null,
        mottaker: {
          id: "123",
          tpe: "FNR",
        },
        relevanteDatoer: [
          {
            dato: "2018-01-01T12:00:00",
            datotype: "DATO_OPPRETTET",
          },
        ],
        dokumenter: [
          {
            dokumentId: "441010176",
            tittel: "MASKERT_FELT",
          },
        ],
      },
      {
        journalpostId: "429108246",
        tittel: "MASKERT_FELT",
        journalposttype: "U",
        journalstatus: "FERDIGSTILT",
        tema: "OPP",
        avsender: null,
        mottaker: {
          id: "123",
          tpe: "FNR",
        },
        relevanteDatoer: [
          {
            dato: "2018-01-01T12:00:00",
            datotype: "DATO_OPPRETTET",
          },
        ],
        dokumenter: [
          {
            dokumentInfoId: "441007131",
            tittel: "MASKERT_FELT",
          },
        ],
      },
      {
        journalpostId: "428965411",
        tittel: "MASKERT_FELT",
        journalposttype: "I",
        journalstatus: "JOURNALFOERT",
        tema: "OPP",
        avsender: {
          id: "123",
          tpe: "FNR",
        },
        mottaker: null,
        relevanteDatoer: [
          {
            dato: "2018-01-01T12:00:00",
            datotype: "DATO_OPPRETTET",
          },
        ],
        dokumenter: [
          {
            dokumentInfoId: "440831549",
            tittel: "MASKERT_FELT",
          },
          {
            dokumentInfoId: "440831548",
            tittel: "MASKERT_FELT",
          },
        ],
      },
    ],
  },
};
