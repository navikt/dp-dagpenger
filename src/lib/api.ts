export default function api(endpoint: string): string {
  if (process.env.NODE_ENV == "test") {
    const url = new URL(location.href);
    url.pathname = endpoint;
    return url.toString();
  }
  return `${process.env.NEXT_PUBLIC_BASE_PATH}/api${
    endpoint.startsWith("/") ? "" : "/"
  }${endpoint}`;
}
