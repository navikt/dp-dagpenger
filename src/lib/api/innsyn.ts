import { v4 as uuid } from "uuid";

export async function fetchInnsynAPI(
  token: Promise<string>,
  endpoint: string
): Promise<any> {
  const callId = uuid();

  console.log(`callId (${callId}) - Henter ${endpoint} fra innsyn.`);

  return fetch(`${process.env.INNSYN_API}/${endpoint}`, {
    method: "get",
    headers: { Authorization: `Bearer ${await token}` },
  })
    .then(async (res) => {
      const contentType = res.headers.get("content-type");
      if (!contentType || !contentType.includes("application/json")) {
        console.log(res.status, res.statusText, res.headers);
        throw new TypeError(
          `Fikk ikke JSON fra innsyn. Body: (${await res.text()}).`
        );
      }
      console.log(`callId (${callId}) - Fikk svar fra innsyn`);
      return res;
    })
    .then((data) => data.json())
    .catch((err) =>
      console.error(
        `callId (${callId}) - Kunne ikke hente ${endpoint} fra innsyn. Feilmelding: ${err}`
      )
    );
}
