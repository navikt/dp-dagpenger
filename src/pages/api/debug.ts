import { getSession } from "../../auth/hooks/session";
import { NextApiRequest, NextApiResponse } from "next";
import nc from "next-connect";
import tokenx from "../../auth/middlewares/tokenx";
import { AuthedNextApiRequest } from "../../auth/middlewares";

export default nc()
  .use(tokenx)
  .get(async (req: AuthedNextApiRequest, res: NextApiResponse) => {
    const session = await getSession({ req });
    if (!session) return res.status(401).end();
    const token = await req.getToken(
      session.user.id_token,
      "dev-gcp:teamdagpenger:dp-innsyn"
    );
    const data = await fetch(process.env.INNSYN_API, {
      method: "get",
      headers: { Authorization: `Bearer ${token}` },
    }).then((data) => data.json());

    res.json(data);
  });
