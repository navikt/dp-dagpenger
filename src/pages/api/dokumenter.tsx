import { NextApiResponse } from "next";
import { gql, GraphQLClient } from "graphql-request";
import { v4 as uuidv4 } from "uuid";
import { AuthedNextApiRequest, withMiddleware } from "../../auth/middlewares";
import { Query } from "../../saf";

const endpoint = "https://safselvbetjening.dev-fss-pub.nais.io/graphql";

export type Dokument = {
  journalpostId: string;
  tittel: string;
};

async function hentDokumenter(
  token: string,
  fnr: string
): Promise<Pick<Query, "dokumentoversiktSelvbetjening">> {
  const callId = uuidv4();
  const variables = { fnr };

  const query = gql`
    query dokumentoversiktSelvbetjening($fnr: String!) {
      dokumentoversiktSelvbetjening(ident: $fnr, tema: [DAG, OPP]) {
        tema {
          journalposter {
            journalpostId
            tittel
            journalposttype
            journalstatus
            dokumenter {
              dokumentInfoId
              tittel
            }
          }
        }
      }
    }
  `;

  const client = new GraphQLClient(endpoint, {
    headers: {
      Authorization: `Bearer ${token}`,
      "Nav-Callid": callId,
      "Nav-Consumer-Id": "dp-dagpenger",
    },
  });

  try {
    console.log(`Henter dokumenter med call-id: ${callId}`);
    return await client.request(query, variables);
  } catch (error) {
    console.error(`Feil fra SAF med call-id ${callId}: ${error}`);
    throw error;
  }
}

export async function handleDokumenter(
  req: AuthedNextApiRequest,
  res: NextApiResponse<Dokument[]>
) {
  const user = req.user;
  if (!user) return res.status(401).end();
  const token = await user.tokenFor(
    "dev-fss:teamdokumenthandtering:safselvbetjening"
  );

  let journalposter;
  try {
    const {
      dokumentoversiktSelvbetjening: { tema },
    } = await hentDokumenter(token, user.fnr);
    journalposter = tema[0].journalposter;
  } catch (errors) {
    return res.status(500).send(errors);
  }

  const dokumenter = journalposter.map(({ journalpostId, tittel }) => ({
    id: journalpostId,
    tittel,
  }));

  res.status(200).json(dokumenter);
}

export default withMiddleware(handleDokumenter);
