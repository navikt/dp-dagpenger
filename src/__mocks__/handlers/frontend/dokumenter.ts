import { faker } from "@faker-js/faker";
import { Journalpost } from "../../../pages/api/dokumenter";
import syntheticUserFnr from "./syntheticUserFnr";
import { AvsenderMottakerIdType, Journalposttype } from "../../../saf";

export const dokumentListeResolver = (req, res, ctx) => {
  const journalposter: Journalpost[] = [...Array(12)].map((_, ji) => {
    const journalpostId = faker.number.toString();
    const antallDokumenter = ji === 4 ? 1 : 3;
    return {
      journalpostId,
      tittel: ji === 0 ? "" : faker.lorem.sentence(),
      journalposttype: Journalposttype.U,
      dato: faker.date.past().toISOString(),
      brukerErAvsenderMottaker: ji === 2 ? true : false,
      avsender: null,
      mottaker: {
        id: ji === 2 ? syntheticUserFnr : faker.string.uuid(),
        type: AvsenderMottakerIdType.Fnr,
      },
      tema: "DAG",
      dokumenter: [...Array(antallDokumenter)].map((_, i) => {
        const id = faker.string.uuid();
        const type = i == 0 ? "Hoved" : "Vedlegg";
        const skjulDokument = (ji === 3 && i === 0) || (ji === 2 && i === 1);
        return {
          id,
          tittel: faker.lorem.sentence(),
          type,
          brukerHarTilgang: !skjulDokument,
          links: [
            {
              href: `/api/dokumenter/${journalpostId}/${id}/forhandsvisning`,
              rel: "preview",
              type: "GET",
            },
          ],
        };
      }),
    };
  });
  return res(ctx.json(journalposter));
};
