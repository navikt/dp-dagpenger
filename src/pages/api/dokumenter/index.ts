import { NextApiHandler } from "next";
import { AvsenderMottaker, Datotype, Journalposttype } from "../../../saf";
import { hentDokumentOversikt } from "../../../lib/saf.service";
import { withSentry } from "@sentry/nextjs";
import { getSession } from "../../../lib/auth.utils";
import { decodeJwt } from "@navikt/dp-auth";
import { safAudience } from "../../../lib/audience";
import { logger } from "@navikt/next-logger";

export type Journalpost = {
  journalpostId: string;
  journalposttype: Journalposttype;
  tittel: string;
  dato: string;
  tema: string;
  dokumenter: Dokument[];
  avsender: AvsenderMottaker;
  mottaker: AvsenderMottaker;
  brukerErAvsenderMottaker: boolean;
};
export type Dokument = {
  id: string;
  tittel: string;
  links: Link[];
  type: DokumentType;
  brukerHarTilgang: boolean;
};
export type DokumentType = "Hoved" | "Vedlegg";
export type Link = { href: string; rel: LinkRel; type: LinkType };
export type LinkType = "GET" | "POST";
export type LinkRel = "preview";

export const handleDokumenter: NextApiHandler<Journalpost[]> = async (
  req,
  res
) => {
  const session = await getSession(req);
  if (!session) return res.status(401).end();
  const payload = decodeJwt(session.token);

  const fnr = payload.pid as string;

  let jposter;
  try {
    const {
      dokumentoversiktSelvbetjening: { journalposter },
    } = await hentDokumentOversikt(await session.apiToken(safAudience), fnr);
    jposter = journalposter;

    logger.info(`Hentet ${jposter.length} journalposter`);
  } catch (errors) {
    logger.error(`Feil fra SAF: ${errors.response}`);
    return res.status(500).end();
  }

  const mapTilRettDato = ({ relevanteDatoer, ...rest }) => {
    const { dato } = relevanteDatoer.find(
      (dato) => dato.datotype == Datotype.DatoOpprettet
    );
    return {
      dato,
      ...rest,
    };
  };

  const berikAvsenderMottaker = ({ avsender, mottaker, ...rest }) => {
    const brukerEr = (am: AvsenderMottaker) =>
      am.type == "FNR" && am.id === fnr;

    const brukerErAvsenderEllerMottaker = () => {
      if (avsender) return brukerEr(avsender);
      if (mottaker) return brukerEr(mottaker);
      return false;
    };
    return {
      brukerErAvsenderMottaker: brukerErAvsenderEllerMottaker(),
      ...rest,
    };
  };

  const journalpostRespons: Journalpost[] = jposter
    .map(mapTilRettDato)
    .map(berikAvsenderMottaker)
    .map(({ journalpostId, dokumenter, ...rest }) => {
      const berikDokmedType = (dok, index) => ({
        // FÃ¸rste vedlegg er alltid hoveddokument
        type: index == 0 ? "Hoved" : "Vedlegg",
        ...dok,
      });

      const berikMedBrukerTilgang = ({ dokumentvarianter, ...rest }) => {
        const erArkiv = (variant) => variant.variantformat === "ARKIV";
        const getArkivVariant = (dokVarianter) => {
          if (dokVarianter) return dokVarianter.find(erArkiv);
          return null;
        };

        const dokumentetKanVises = () => {
          const variant = getArkivVariant(dokumentvarianter);
          if (variant) return variant.brukerHarTilgang;
          return false;
        };
        return { brukerHarTilgang: dokumentetKanVises(), ...rest };
      };

      const byggForhandsvisningLink = ({ dokumentInfoId, ...rest }) => ({
        links: [
          {
            href: `${process.env.NEXT_PUBLIC_BASE_PATH}/api/dokumenter/${journalpostId}/${dokumentInfoId}/forhandsvisning`,
            rel: "preview",
            type: "GET",
          },
        ],
        dokumentInfoId,
        ...rest,
      });

      return {
        journalpostId,
        ...rest,
        dokumenter: dokumenter
          .map(berikDokmedType)
          .map(berikMedBrukerTilgang)
          .map(byggForhandsvisningLink)
          .map(({ dokumentInfoId, ...rest }) => ({
            id: dokumentInfoId,
            ...rest,
          })),
      };
    });

  res.json(journalpostRespons);
};

export default withSentry(handleDokumenter);
