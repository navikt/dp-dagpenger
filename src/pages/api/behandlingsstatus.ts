import { NextApiHandler } from "next";
import { withSentry } from "@sentry/nextjs";
import { getSession } from "../../lib/auth.utils";
import { fetchInnsynAPI } from "../../lib/api/innsyn";

const antallDager = 28;

export type Behandlingsstatus = {
  status: Status;
  antallSøknader: number;
  antallVedtak: number;
};

export type Status =
  | null
  | "UnderBehandling"
  | "FerdigBehandlet"
  | "UnderOgFerdigBehandlet";

export async function hentBehandlingsstatus(
  token: Promise<string>
): Promise<Behandlingsstatus> {
  const fom = new Date();
  fom.setDate(fom.getDate() - antallDager);

  const [søknader, vedtak] = await Promise.all([
    fetchInnsynAPI(token, `soknad?soktFom=${getISODate(fom)}`),
    fetchInnsynAPI(token, `vedtak?fattetFom=${getISODate(fom)}`),
  ]);

  if (!søknader || !vedtak) {
    // Kunne ikke hente søknader og/eller vedtak, så vi viser ikke status
    return { status: null, antallSøknader: 0, antallVedtak: 0 };
  }

  const antallSøknader = søknader.length;
  const antallVedtak = vedtak.length;

  let status: Status = null;
  if (antallSøknader > antallVedtak && antallVedtak > 0) {
    status = "UnderOgFerdigBehandlet";
  } else if (antallVedtak > 0) {
    status = "FerdigBehandlet";
  } else if (antallSøknader > 0) {
    status = "UnderBehandling";
  }

  return {
    status,
    antallSøknader,
    antallVedtak,
  };
}

export const handleBehandlingsstatus: NextApiHandler<Behandlingsstatus> =
  async (req, res) => {
    const { token, apiToken } = await getSession(req);
    if (!token) return res.status(401).end();

    const audience = `${process.env.NAIS_CLUSTER_NAME}:teamdagpenger:dp-innsyn`;

    return hentBehandlingsstatus(apiToken(audience)).then(res.json);
  };

export default withSentry(handleBehandlingsstatus);

function getISODate(date: Date) {
  return (
    date.getFullYear() +
    "-" +
    ("0" + (date.getMonth() + 1)).slice(-2) +
    "-" +
    ("0" + date.getDate()).slice(-2)
  );
}
