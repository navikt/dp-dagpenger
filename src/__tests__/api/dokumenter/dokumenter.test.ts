import { createMocks } from "node-mocks-http";
import { handleDokumenter } from "../../../pages/api/dokumenter";

describe("/api/dokumenter", () => {
  test("svarer med en liste dokumenter", async () => {
    const { req, res } = createMocks({
      method: "GET",
      user: {
        fnr: "123123123",
        tokenset: {
          access_token: "123",
        },
        tokenFor: (token) => token,
      },
    });

    // @ts-ignore MockRequest matcher ikke AuthedNextApiRequest
    await handleDokumenter(req, res);

    expect(res._getStatusCode()).toBe(200);
    expect(res._getJSONData()).toMatchSnapshot();
  });
});
