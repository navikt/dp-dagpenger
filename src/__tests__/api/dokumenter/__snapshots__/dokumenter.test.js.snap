// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/dokumenter svarer med en liste dokumenter 1`] = `
Array [
  Object {
    "dokumenter": Array [
      Object {
        "tittel": "MASKERT_FELT",
      },
    ],
    "id": "429111291",
    "tema": "DAG",
    "tittel": "MASKERT_FELT",
  },
  Object {
    "dokumenter": Array [
      Object {
        "id": "441007131",
        "tittel": "MASKERT_FELT",
      },
    ],
    "id": "429108246",
    "tema": "DAG",
    "tittel": "MASKERT_FELT",
  },
  Object {
    "dokumenter": Array [
      Object {
        "id": "440831549",
        "tittel": "MASKERT_FELT",
      },
      Object {
        "id": "440831548",
        "tittel": "MASKERT_FELT",
      },
    ],
    "id": "428965411",
    "tema": "DAG",
    "tittel": "MASKERT_FELT",
  },
  Object {
    "dokumenter": Array [
      Object {
        "tittel": "MASKERT_FELT",
      },
    ],
    "id": "429111291",
    "tema": "OPP",
    "tittel": "MASKERT_FELT",
  },
  Object {
    "dokumenter": Array [
      Object {
        "id": "441007131",
        "tittel": "MASKERT_FELT",
      },
    ],
    "id": "429108246",
    "tema": "OPP",
    "tittel": "MASKERT_FELT",
  },
  Object {
    "dokumenter": Array [
      Object {
        "id": "440831549",
        "tittel": "MASKERT_FELT",
      },
      Object {
        "id": "440831548",
        "tittel": "MASKERT_FELT",
      },
    ],
    "id": "428965411",
    "tema": "OPP",
    "tittel": "MASKERT_FELT",
  },
]
`;
